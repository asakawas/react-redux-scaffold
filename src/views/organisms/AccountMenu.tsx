import React, { FunctionComponent, useState, useCallback } from 'react';
import ExitToAppIcon from '@material-ui/icons/ExitToApp';
import Divider from '@material-ui/core/Divider';
import Menu from '@material-ui/core/Menu';
import Typography from '@material-ui/core/Typography';
import AccountMenuButton from '~/views/atoms/buttons/AccountMenuButton';
import PopupMenuItem, { PopupMenuItemProps } from '~/views/molecules/PopupMenuItem';
import { User } from '~/state/ducks/user/models';

/**
 * The type of props of AccountMenu.
 */
type AccountMenuProps = Readonly<{
  /** The event handler called when the menu item 'Sign Out' is clicked. */
  onSignOutClick: NonNullable<PopupMenuItemProps['onClick']>;

  /** Current user. */
  currentUser: User;
}>;

const AccountMenu: FunctionComponent<AccountMenuProps> = ({ onSignOutClick, currentUser }) => {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);

  const handleAccountMenuOpen = useCallback((event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  }, []);
  const handleAccountMenuClose = useCallback(() => {
    setAnchorEl(null);
  }, []);

  const handleSignOutClick = useCallback(
    (event: React.MouseEvent<HTMLLIElement, MouseEvent>) => {
      setAnchorEl(null);
      onSignOutClick(event);
    },
    [onSignOutClick],
  );

  return (
    <>
      <AccountMenuButton onAccountMenuButtonClick={handleAccountMenuOpen} />
      <Menu
        aria-label="account menu"
        anchorEl={anchorEl}
        anchorOrigin={{
          vertical: 'top',
          horizontal: 'right',
        }}
        keepMounted
        transformOrigin={{
          vertical: 'top',
          horizontal: 'right',
        }}
        open={open}
        onClose={handleAccountMenuClose}
      >
        <Typography align="center" noWrap gutterBottom>
          {currentUser.name}
        </Typography>
        <Divider />
        <PopupMenuItem text="Sign Out" icon={<ExitToAppIcon />} onClick={handleSignOutClick} />
      </Menu>
    </>
  );
};

export default React.memo(AccountMenu);
