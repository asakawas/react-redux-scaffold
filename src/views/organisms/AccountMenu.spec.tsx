import React from 'react';
import { render, wait, fireEvent, screen } from '@testing-library/react';
import AccountMenu from './AccountMenu';

describe('AccountMenu', () => {
  test('renders the menu button as expected', () => {
    const { asFragment } = render(
      <AccountMenu onSignOutClick={() => {}} currentUser={{ name: 'hoge' }} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders the menu as expected', () => {
    render(<AccountMenu onSignOutClick={() => {}} currentUser={{ name: 'hoge' }} />);
    expect(screen.getByLabelText('account menu')).toMatchSnapshot();
  });

  test("shows the popup menu only after it's clicked", () => {
    const renderResult = render(
      <AccountMenu onSignOutClick={() => {}} currentUser={{ name: 'hoge' }} />,
    );

    const accountMenu = screen.getByLabelText('account menu');
    expect(accountMenu).not.toBeVisible();

    const accountMenuButton = renderResult.getByLabelText('account menu button');
    fireEvent.click(accountMenuButton);
    expect(accountMenu).toBeVisible();
  });

  test('closes the popup menu when a menu item is clicked', async () => {
    const renderResult = render(
      <AccountMenu onSignOutClick={() => {}} currentUser={{ name: 'hoge' }} />,
    );

    const accountMenuButton = renderResult.getByLabelText('account menu button');
    fireEvent.click(accountMenuButton);

    const accountMenu = screen.getByLabelText('account menu');
    expect(accountMenu).toBeVisible();

    const signOut = screen.getByText('Sign Out');
    fireEvent.click(signOut);
    await wait(() => expect(accountMenu).not.toBeVisible());
  });

  test('closes the popup menu when outside of the menu clicked', async () => {
    const renderResult = render(
      <AccountMenu onSignOutClick={() => {}} currentUser={{ name: 'hoge' }} />,
    );

    const accountMenuButton = renderResult.getByLabelText('account menu button');
    fireEvent.click(accountMenuButton);

    const accountMenu = screen.getByLabelText('account menu');
    expect(accountMenu).toBeVisible();

    // accountMenu and its first child covers whole screen
    // to trap click and key down events.
    fireEvent.click(accountMenu.firstChild);
    await wait(() => expect(accountMenu).not.toBeVisible());
  });

  test("calls the passed handler when 'Sign Out' clicked", () => {
    const handler = jest.fn();
    const renderResult = render(
      <AccountMenu onSignOutClick={handler} currentUser={{ name: 'hoge' }} />,
    );

    const accountMenuButton = renderResult.getByLabelText('account menu button');
    fireEvent.click(accountMenuButton);
    expect(handler).toHaveBeenCalledTimes(0);

    const signOut = screen.getByText('Sign Out');
    fireEvent.click(signOut);
    expect(handler).toHaveBeenCalledTimes(1);
  });
});
