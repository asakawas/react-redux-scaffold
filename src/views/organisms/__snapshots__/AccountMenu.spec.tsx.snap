// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountMenu renders the menu as expected 1`] = `
<div
  aria-hidden="true"
  aria-label="account menu"
  class="MuiPopover-root"
  role="presentation"
  style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px; visibility: hidden;"
>
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiPaper-root MuiMenu-paper MuiPopover-paper MuiPaper-elevation8 MuiPaper-rounded"
    style="opacity: 0; transform: scale(0.75, 0.5625); visibility: hidden;"
    tabindex="-1"
  >
    <ul
      class="MuiList-root MuiMenu-list MuiList-padding"
      role="menu"
      tabindex="-1"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap MuiTypography-gutterBottom MuiTypography-alignCenter"
        tabindex="0"
      >
        hoge
      </p>
      <hr
        class="MuiDivider-root"
      />
      <li
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
        role="menuitem"
        tabindex="-1"
      >
        <div
          class="MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </div>
        <span
          class="MuiTypography-root"
        >
          Sign Out
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </li>
    </ul>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;

exports[`AccountMenu renders the menu button as expected 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #b3d6a4;
}

<button
    aria-label="account menu button"
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiIconButton-edgeEnd"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiIconButton-label"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circle c0 MuiAvatar-colorDefault"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorAction"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
          />
        </svg>
      </div>
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
